### PR TITLE
Support HDR/LDR Light Sources

### PR DESCRIPTION
**Pull request by pandamicro**
Replacing https://github.com/cocos-creator/engine/pull/9433

Native engine: https://github.com/cocos-creator/engine-native/pull/3951

API naming:

1. DirectionalLight
    - illuminance (will be deprecated)
    - illuminanceHDR
    - illuminanceLDR
2. DirectionalLightComponent
    - illuminance
3. SpotLight & SphereLight
    - luminance
    - luminanceHDR
    - luminanceLDR
4. SpotLightComponent & SphereLightComponent
    - luminousFlux
    - luminance

### CODE REVIEW

File: cocos/core/scene-graph/scene-globals.ts [hunk 1]
```diff
@@ -28,21 +28,30 @@ import { ccclass, visible, type, displayOrder, readOnly, slide, range, rangeStep
...
+// Normalize HDR color
+const normalizeHDRColor = (color : Vec4) => {
+    const intensity = 1.0 / Math.max(Math.max(color[0], color[1]), color[2]);
+    if (intensity < 1.0) {
+        for (let i = 0; i < 3; ++i) {
+            color[i] *= intensity;
+        }
+    }
+}
+const clampColor = (x: number) => Math.min(x * 255, 255);
```
Comment (pandamicro) [inline cocos/core/scene-graph/scene-globals.ts]: Make `normalizeHDRColor` and `clampColor` as script level internal tool functions, if no outside classes need this kind of tool functions, then we can use this kind of declaration.
If other classes also need them, we need to put them into tool or utils module

File: cocos/core/scene-graph/scene-globals.ts [hunk 2]
```diff
@@ -89,60 +98,35 @@ export class AmbientInfo {
...
         const colorRef = isHDR ? this._skyColor : this._skyColorLDR;
-
-        const clampColor = (x: number) => Math.min(x * 255, 255);
         return new Color(clampColor(colorRef.x), clampColor(colorRef.y), clampColor(colorRef.z), 255);
     }
 
     set skyColorValue (val: Vec4) {
-        const isHDR = (legacyCC.director.root as Root).pipeline.pipelineSceneData.isHDR;
-        const clampColor = (x: number) => Math.min(x * 255, 255);
-        let colorRef = isHDR ? this._skyColor : this._skyColorLDR;
```
Comment (pandamicro) [inline cocos/core/scene-graph/scene-globals.ts]: Remove useless logic

File: cocos/3d/lights/sphere-light-component.ts [hunk 3]
```diff
@@ -42,7 +42,8 @@ export class SphereLight extends Light {
     @serializable
     protected _size = 0.15;
     @serializable
-    protected _luminance = 1700 / scene.nt2lm(0.15);
+    @formerlySerializedAs('_luminance')
+    protected _luminanceHDR = 1700 / scene.nt2lm(0.15);
```
Comment (pandamicro) [inline cocos/3d/lights/sphere-light-component.ts:46]: This way we can upgrade previous serialized data in `_luminance` to `_luminanceHDR`

File: cocos/core/renderer/scene/directional-light.ts [hunk 4]
```diff
@@ -35,7 +35,7 @@ const _v3 = new Vec3();
 
 export class DirectionalLight extends Light {
     protected _dir: Vec3 = new Vec3(1.0, -1.0, -1.0);
-    protected _illuminance: number = Ambient.SUN_ILLUM;
+    protected _illuminanceHDR: number = Ambient.SUN_ILLUM;
```
Comment (pandamicro) [inline cocos/core/renderer/scene/directional-light.ts:38]: This is internal property, we can refactor directly as long as we change everywhere it's been used

File: cocos/core/renderer/scene/directional-light.ts [hunk 5]
```diff
@@ -50,40 +50,40 @@ export class DirectionalLight extends Light {
...
-    set illuminanceHDR (illum: number) {
-        this._illuminance = illum;
+    get illuminanceHDR () {
+        return this._illuminanceHDR;
+    }
+    set illuminanceHDR (value: number) {
+        this._illuminanceHDR = value;
         if (JSB) {
-            (this._nativeObj as NativeDirectionalLight).setIlluminance(illum);
+            (this._nativeObj as NativeDirectionalLight).setIlluminanceHDR(value);
```
Comment (pandamicro) [inline cocos/core/renderer/scene/directional-light.ts:76]: I also synced native APIs in another pull request linked

File: cocos/core/renderer/scene/directional-light.ts [hunk 6]
```diff
@@ -50,40 +50,40 @@ export class DirectionalLight extends Light {
...
         }
     }
 
-    set illuminanceLDR (illum: number) {
-        this._illuminanceLDR = illum;
+    get illuminanceLDR () {
+        return this._illuminanceLDR;
+    }
+    set illuminanceLDR (value: number) {
```
Comment (pandamicro) [inline cocos/core/renderer/scene/directional-light.ts:83]: `illum` is not a word, it's preferred to use correct wording to improve readability

File: cocos/core/pipeline/forward/forward-stage.ts [hunk 7]
```diff
@@ -169,17 +169,10 @@ export class ForwardStage extends RenderStage {
...
-                colors[0].z *= scale;
-            } else {
-                colors[0].x = camera.clearColor.x;
-                colors[0].y = camera.clearColor.y;
-                colors[0].z = camera.clearColor.z;
-            }
+            SRGBToLinear(colors[0], camera.clearColor);
+            colors[0].x = camera.clearColor.x;
+            colors[0].y = camera.clearColor.y;
+            colors[0].z = camera.clearColor.z;
```
Comment (minggo) [inline cocos/core/pipeline/forward/forward-stage.ts]: The logic is strange. colors[0] is assigned twice. SRGBToLinear() is not needed or 
```javascript
colors[0].x = camera.clearColor.x;
colors[0].y = camera.clearColor.y;
colors[0].z = camera.clearColor.z;
```
is not needed.

And in native codes, only `SRGBToLinear() ` is invoked.

File: cocos/core/pipeline/pipeline-scene-data.ts [hunk 8]
```diff
@@ -74,15 +74,6 @@ export class PipelineSceneData {
...
     }
-    public get fpScale () {
-        return this._fpScale;
-    }
-    public set fpScale (val: number) {
-        this._fpScale = val;
-        if (JSB) {
-            this._fpScale = val;
-        }
-    }
```
Comment (minggo) [inline cocos/core/pipeline/pipeline-scene-data.ts:85]: should delete the element in native codes too: https://github.com/cocos-creator/engine-native/blob/v3.4.0/cocos/scene/Define.h#L110

File: cocos/core/pipeline/pipeline-ubo.ts [hunk 9]
```diff
@@ -102,23 +101,29 @@ export class PipelineUBO {
...
-        cv.set(skyColor, UBOCamera.AMBIENT_SKY_OFFSET);
-        cv.set(ambient.albedoArray, UBOCamera.AMBIENT_GROUND_OFFSET);
+        cv[UBOCamera.AMBIENT_SKY_OFFSET + 0] = skyColor.x;
+        cv[UBOCamera.AMBIENT_SKY_OFFSET + 1] = skyColor.y;
+        cv[UBOCamera.AMBIENT_SKY_OFFSET + 2] = skyColor.z;
+        cv[UBOCamera.AMBIENT_SKY_OFFSET + 3] = skyColor.w;
+        cv[UBOCamera.AMBIENT_GROUND_OFFSET + 0] = ambient.groundAlbedo.x;
+        cv[UBOCamera.AMBIENT_GROUND_OFFSET + 1] = ambient.groundAlbedo.y;
+        cv[UBOCamera.AMBIENT_GROUND_OFFSET + 2] = ambient.groundAlbedo.z;
+        cv[UBOCamera.AMBIENT_GROUND_OFFSET + 3] = ambient.groundAlbedo.w;
```
Comment (minggo) [inline cocos/core/pipeline/pipeline-ubo.ts:126]: It seems the logic is different with native

File: cocos/core/renderer/scene/skybox.ts [hunk 10]
```diff
@@ -59,8 +59,22 @@ export class Skybox {
...
+     * @zh 是否启用HDR？
+     */
+    get useHDR (): boolean {
+        return this._useHDR;
+    }
+
+    set useHDR (val: boolean) {
+        this._setUseHDR(val);
+        const tex = this.envmap;
+        this.envmap = tex;
```
Comment (minggo) [inline cocos/core/renderer/scene/skybox.ts]: The logic is strange, why do it like this?

File: cocos/core/renderer/scene/skybox.ts [hunk 11]
```diff
@@ -142,23 +203,43 @@ export class Skybox {
         }
     }
 
-    private _setIsRGBE (val) {
-        this._isRGBE = val;
+    private _setUseHDR (val) {
+        this._useHDR = val;
         if (JSB) {
-            this._nativeObj!.isRGBE = val;
+            this._nativeObj!.useHDR = val;
```
Comment (minggo) [inline cocos/core/renderer/scene/skybox.ts:196]: should change the element name in native too https://github.com/cocos-creator/engine-native/blob/v3.4.0/cocos/scene/Define.h#L108

File: cocos/core/renderer/scene/skybox.ts [hunk 12]
```diff
@@ -142,23 +203,43 @@ export class Skybox {
...
+        }
+    }
+
+    private _setUseDiffuseMap (val) {
+        this._useDiffuseMap = val;
+        if (JSB) {
+            this._nativeObj!.useDiffuseMap = val;
```
Comment (minggo) [inline cocos/core/renderer/scene/skybox.ts:203]: should add an element in native codes https://github.com/cocos-creator/engine-native/blob/v3.4.0/cocos/scene/Define.h#L93

**Review by minggo — APPROVED** [review]

**Pull request merged**